ToolBox/GDAXDownloader/GDAXDownloader.cs

```diff
@@ -41,77 +47,101 @@ public class GDAXDownloader : IDataDownloader
-                                    Close = decimal.Parse(barValues[4].ToString()),
-                                    Volume = decimal.Parse(barValues[5].ToString()),
-                                    Value = decimal.Parse(barValues[4].ToString()),
-                                    DataType = MarketDataType.TradeBar,
-                                    Period = Time.OneHour,
-                                    EndTime = dt.AddHours(1)
-                                };
-                                returnData.Add(tradeBar);
-                            }
-                        }
-                        
+                        Log.Error("REQUEST FAILED: " + request.Address);
+                        throw;
                     }
-                    counter = counter.AddDays(1);
+                    Log.Trace("WARNING: Web request failed with message " + ex.Message + "Retrying... " + retryCount + " times");
                 }
-                catch(System.Net.WebException ex)
+            }
+            return data;
+        }
+
+        /// <summary>
+        /// Parse string response from web response
+        /// </summary>
+        /// <param name="Symbol">Crypto security symbol.</param>
+        /// <param name="granularity">Resolution in seconds.</param>
+        /// <param name="data">Web response as string.</param>
+        /// <returns>web response as string</returns>
+        List<BaseData> ParseCandleData(Symbol symbol, double granularity, string data)
+        {
+            List<BaseData> returnData = new List<BaseData>();
+            if (data.Length > 0)
+            {
+                var parsedData = JsonConvert.DeserializeObject<string[][]>(data);
+
+                foreach (var datapoint in parsedData)
                 {
-                    Log.Error("BAD REQUEST: " + requestURL);
+                    var epochs = double.Parse(datapoint[0].ToString());
+                    var tradeBar = new TradeBar()
+                    {
+                        Time = Time.UnixTimeStampToDateTime(epochs),
+                        Symbol = symbol,
+                        Low = decimal.Parse(datapoint[1].ToString()),
+                        High = decimal.Parse(datapoint[2].ToString()),
+                        Open = decimal.Parse(datapoint[3].ToString()),
+                        Close = decimal.Parse(datapoint[4].ToString()),
+                        Volume = decimal.Parse(datapoint[5].ToString()),
+                        Value = decimal.Parse(datapoint[4].ToString()),
+                        DataType = MarketDataType.TradeBar,
+                        Period = new TimeSpan(0, 0, (int)granularity),
+                        EndTime = Time.UnixTimeStampToDateTime(epochs).AddSeconds(granularity)
+                    };
+                    returnData.Add(tradeBar);
                 }
             }
-            return returnData;
+            return returnData.OrderBy(datapoint => datapoint.Time).ToList();

```

Comment:
Curious did you ever see this coming back in the wrong order?

Reply:
The gdax response returns the data from latest->oldest order, thereby requiring reordering before piping it to the writer.